docs: Quilt docs (#2272)
MystenLabs/walrus: docs/book/usage/quilt.md

blob is stored in a different quilt. Moreover, individual blobs cannot be deleted, extended or shared
separately; these operations can only be applied to the entire quilt.

## Target Use Cases

Using *quilt* requires minimal additional effort beyond standard procedures. The primary
considerations are that the unique ID assigned to each blob within a quilt cannot be
directly derived from its contents, unlike a regular Walrus blob_id, deletion, extension
and share operations are not allowed on individual blobs in the quilt, only the quilt
can be the target.

### Lower Cost

This is the most clear and significant use case. *quilt* is especially advantageous for
managing large volumes of small blobs, as long as they can be grouped together by the
user. By consolidating multiple small blobs into a single quilt, storage costs can be
reduced dramatically—up to 50 times for files around 100KB—making it an efficient
solution for cost-sensitive applications.

The following table demonstrates shows the potential cost savings in WAL when storing 600 small
blobs for 1 epoch as a quilt compared to storing them as separate blobs.

> | Blob size | Regular blob storage cost | Quilt storage cost | Cost saving factor |

Reviewer: Suggestion: make sure it's clear that the cost is not for the quilt, but `quilt_cost / 600`
```suggestion
| Blob size | Regular blob storage cost | Quilt Patch storage cost | Cost saving factor |
```

|----------:|--------------------------:|-------------------:|-------------------:|
|      10KB |                 2.088 WAL |          0.005 WAL |               409x |
|      50KB |                 2.088 WAL |          0.011 WAL |               190x |
|     100KB |                 2.088 WAL |          0.020 WAL |               104x |
|     200KB |                 2.088 WAL |          0.036 WAL |                58x |
|     500KB |                 2.136 WAL |          0.084 WAL |                25x |
|       1MB |                 2.208 WAL |          0.170 WAL |                13x |

> **Note:** The costs shown in this table are for illustrative purposes only. Actual
> costs may vary due to changes in smart contract parameters, network conditions, and
> other factors.

The table only shows the Walrus storage cost. Similar savings are possible for Sui gas fees. Storing
600 blobs as a quilt results in gas fees of 0.00725 SUI compared to 1.725 SUI when storing them as

> standard blobs (batching 25 at a time), which is a savings factor of **238x**. These values are not

Reviewer: Suggestion: "batching 25 at a time" is a bit confusing here. I think you are talking about using the store command with the batched register/certifiy, right? maybe clarify a little?

# Quilt: Walrus Native Batch Store Tool

Quilt is a batch storage feature, aiming to optimize the storage cost and efficiency of large

> numbers of small blobs. Prior to Quilt, storing small blobs (less than 10MB) in Walrus

Reviewer 2: Nit: Can we use lower-case "quilt" except at the start of sentences? If we want to highlight it, I'd use italic font (`*quilt*`).

involved higher per-byte costs due to internal system data overhead. Quilt addresses this by
encoding multiple (up to 666 for QuiltV1) blobs into a single unit called a quilt, significantly
reducing Walrus storage overhead and lowering costs to purchase Walrus/Sui storage, as well as Sui
computation gas fees.

Importantly, each blob within a quilt can be accessed and retrieved individually, without needing

> to download the entire quilt. Moreover, the blob boundaries in a quilt aligns with Walrus

Reviewer 2: Nit:
```suggestion
to download the entire quilt. Moreover, the blob boundaries in a quilt align with Walrus
```

internal structures as well as Walrus storage nodes, this allows for retrieval latency that is
comparable to, or even lower than, that of a regular blob.

In addition, Quilt introduces custom, immutable `Walrus-native` blob metadata, allowing user to
assign different types of `metadata` to each blob in a quilt, for example, unique identifiers, as
well as tags of key-value pairs. Note, this metadata is functionally similar to the existing
Blob Metadata store on-chain, however, there are some fundamental distinctions. First,
`Walrus-native` metadata is stored alongside the blob data, and hence it reduces costs and
simplifies management. Second, this metadata can be used for efficient lookup of blobs within a
quilt, for example, reading blobs with a particular tag. When storing a quilt, users can set

> the `Walrus-native` metadata via the Quilt Apis.

Reviewer 2: Nit:
```suggestion
the `Walrus-native` metadata via the quilt APIs.
```

that differs from the `BlobId` used for regular Walrus blobs, and a `QuiltPatchId` is determined by
the composition of the entire quilt, ranther than the single blob, and hence it may change if the
blob is stored in a different quilt. Moreover, individual blobs cannot be deleted, extended or shared
separately; these operations can only be applied to the entire quilt.

## Target Use Cases

Using Quilt requires minimal additional effort beyond standard procedures. The primary
considerations are that the unique ID assigned to each blob within a quilt cannot be
directly derived from its contents, unlike a regular Walrus blob_id, deletion, extension
and share operations are not allowed on individual blobs in the quilt, only the quilt
can be the target.

### Lower Cost

This is the most clear and significant use case. Quilt is especially advantageous for
managing large volumes of small blobs, as long as they can be grouped together by the
user. By consolidating multiple small blobs into a single quilt, storage costs can be
reduced dramatically—up to 50 times for files around 100KB—making it an efficient
solution for cost-sensitive applications.

> The following table demonstrates the potential cost savings when blobs are stored with
> Quilt:
> 
> > **Note:** The costs shown in this table are for illustrative purposes only. Actual
> > costs may vary due to changes in smart contract parameters, network conditions, and
> > other factors.
> 
> | # | Blob Size | # of Blobs | Regular Blob Cost¹ | Quilt Cost | Cost Saving Factor |
> |---|-----------|------------|-------------------|------------|-------------------|
> | 1 | 10KB | 600 | 1.725 SUI + 2.088 WAL | 0.00725 SUI + 0.0051 WAL | SUI: 238x WAL: 409x |
> | 2 | 50KB | 600 | 1.725 SUI + 2.088 WAL | 0.00725 SUI + 0.011 WAL | SUI: 238x WAL: 190x |
> | 3 | 100KB | 600 | 1.725 SUI + 2.088 WAL | 0.00725 SUI + 0.02 WAL | SUI: 238x WAL: 104x |
> | 4 | 200KB | 600 | 1.869 SUI + 2.088 WAL | 0.00725 SUI + 0.036 WAL | SUI: 238x WAL: 58x |
> | 5 | 500KB | 600 | 2.085 SUI + 2.136 WAL | 0.00725 SUI + 0.084 WAL | SUI: 238x WAL: 25x |
> | 6 | 1MB | 600 | 2.085 SUI + 2.208 WAL | 0.00725 SUI + 0.17 WAL | SUI: 238x WAL: 13x |
> 
> *Table: Cost savings when storing 600 blobs of the same size for 1 epoch, comparing
> the traditional approach (batching 25 blobs at a time) versus using Quilt.*
> 
> ¹ Regular blob costs are calculated based on batching 25 blobs at a time.
> 
> As shown in the table, Quilt provides significant cost savings for small blobs, with
> the greatest benefits for blobs under 100KB where both SUI and WAL costs can be
> reduced by over 50x.

Reviewer 2: A few suggestions for this table:
1. As the number of blobs is the same in all cases, this could be removed and mentioned separately.
2. The SUI cost *should* be independent of the blob size. This could therefore also be removed from the table and mentioned separately.
3. The first column isn't necessary.
4. Right-align the columns.
5. Nit: Reformat the table to align the columns. (There is a useful VSCode extension for this: `darkriszty.markdown-table-prettify`.)

Overall, something like the following:
```suggestion
The following table demonstrates shows the potential cost savings in WAL when storing 600 small
blobs for 1 epoch as a quilt compared to storing them as separate blobs.

| Blob size | Regular blob storage cost | Quilt storage cost | Cost saving factor |
|----------:|--------------------------:|-------------------:|-------------------:|
|      10KB |                 2.088 WAL |          0.005 WAL |               409x |
|      50KB |                 2.088 WAL |          0.011 WAL |               190x |
|     100KB |                 2.088 WAL |          0.020 WAL |               104x |
|     200KB |                 2.088 WAL |          0.036 WAL |                58x |
|     500KB |                 2.136 WAL |          0.084 WAL |                25x |
|       1MB |                 2.208 WAL |          0.170 WAL |                13x |

> **Note:** The costs shown in this table are for illustrative purposes only. Actual
> costs may vary due to changes in smart contract parameters, network conditions, and
> other factors.

The table only shows the Walrus storage cost. Similar savings are possible for Sui gas fees. Storing
600 blobs as a quilt results in gas fees of 0.00725 SUI compared to 1.725 SUI when storing them as
standard blobs (batching 25 at a time), which is a savings factor of 238x. These values are not
affected by the blob size.

Overall, quilt provides significant cost savings for small blobs, with the greatest benefits for
blobs under 100KB where both SUI and WAL costs can be reduced by over 50x.
```
